### PR TITLE
Support 3D Custom Switch Blocks 

### PR DESCRIPTION
**Pull request by saaramahmoudi**
#### Description

This pull request support 3d custom switch blocks in the architecture file and automatically add the inter-die edges between tracks in different layer to RR graph. 

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by vaughnbetz** [comment]
Please attach VTR, Titan and Koios QoR data to make sure we haven't degraded it.

File: vpr/src/route/rr_graph.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph.cpp:1265]: Needs a big comment defining dimensions so we can see how to access. E.g. [0..grid.layers-1][0..grid.width-1][grid.height-1][0..max_chan_width-1][CHANX or CHANY].

File: vpr/src/route/rr_graph.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph.cpp:1269]: Comment: need a 3D custom switchblock.

File: vpr/src/route/rr_graph.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph.cpp:1382]: Is this a change in the master?

File: vpr/src/route/rr_graph.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph.cpp:2160]: connection -> connections
explain what the newly added CHANX nodes are.

File: vpr/src/route/rr_graph.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph.cpp:3272]: Expand this comment explaining what is going on.  Nodes have already been created in the lookup so you're just going through the lookup until you run out of nodes; you create all these extra nodes. The extra nodes are ...

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:96]: Comment

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:114]: Comment

File: vpr/src/route/rr_graph2.h
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.h:32]: Comment.

File: vpr/src/route/rr_graph2.h
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.h:72]: Comment more. How do I access the returned 2D matrix?  What do I pass in (that part can be brief).

File: vpr/src/route/rr_graph2.h
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.h:38]: Comment.

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:1158]: Change to NUM_2D_SIDES and NUM_3D_SIDES

Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:1158]: Could make a little function to check if you are crossing a layer based on the sides (easier to read, doesn't embed the numbering of sides as deeply in the logic).

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:1174]: I'd have a variable chany_index = 1 and chanx_index = 0 to make this easier to read.
May be some code duplication that a helper function could remove.

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:1214]: "could have used either CHANX or CHANY; we chose to use all CHANX"

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:1219]: upper limit should have -1

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:1224]: require global resource routing -> have inter-cluster routing

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:2215]: May even want ASCII art here.

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:2267]: Should move to declaration.

**Review by vaughnbetz — CHANGES_REQUESTED** [review]
1. Comments :).
2. Don't need the vector of channel types (just make the end user specify more in the custom sb edges if they want multiple sides).
3. ASCII art of how the sb spec turns into 2 length 0 wires and some edges. Also only works with muxes: comment that.
4. FUTURE WORK: should allow muxing onto the 3D wire (done) and fanout from the 3D wire to the destinations on the other layer (not done).

**Comment by vaughnbetz** [comment]
Another possible future work (maybe fits in the same time): bidirectional wires.

File: vpr/src/route/build_switchblocks.h
Comment (saaramahmoudi) [reply vpr/src/route/build_switchblocks.h:81]: If the switch index indicates an inter-die connection and we have more than two layers, connection can be going to the above die or the below die. Without the layer index for from and to wire, there is no possible way to handle that.

File: vpr/src/route/build_switchblocks.cpp
Comment (saaramahmoudi) [inline vpr/src/route/build_switchblocks.cpp:990]: should change to true

**Comment by saaramahmoudi** [comment]
@vaughnbetz Here is QoR check for VTR and Titan benchmarks for 2D FPGAs.
[QoR check](https://docs.google.com/spreadsheets/d/1POATTw4VJ54p5i5eTEh13EjBtQw9ohvOKS8PwCIVrHc/edit?usp=sharingl)

**Comment by vaughnbetz** [comment]
Thanks. Please also do a comparison of maximum memory footprint.
There are some CI QoR failures that are due to an increase in maximum memory.  nightly_test3 has some for example with a complex switch block (up over 20% vs. golden results).

**Comment by saaramahmoudi** [comment]
@vaughnbetz 
[Titan23](https://docs.google.com/spreadsheets/d/1rPlzp8oUOkieDYCrdwDivxKVaUHDMu-MKkYa95soiJo/edit?usp=sharing) and [VTR](https://docs.google.com/spreadsheets/d/1zV_jwlpJ4x88APbUACVDKTqpsW4GlMM64IPTi8c_QEg/edit?usp=sharing) show the qor_compare script outputs for these benchmarks. 3D custom SBs have some max memory increase for both, 4% and 2.5%. I can dig into it more if we can't merge this with max memory increase.

File: vpr/src/base/read_options.cpp
Comment (vaughnbetz) [inline vpr/src/base/read_options.cpp:2464]: This seems like it would fit better in the arch file (as a new parameter) than as a command line parameter.  What do you think?

File: vpr/src/route/rr_graph2.cpp
Comment (vaughnbetz) [inline vpr/src/route/rr_graph2.cpp:2340]: Is there a big comment somewhere saying how custom_3d_sb_fanin_fanout works?  It should detail exactly what the rr graph this generates looks like.

File: vpr/src/base/read_options.cpp
Comment (saaramahmoudi) [reply vpr/src/base/read_options.cpp:2464]: I also think that it might be better to put it in the architecture file, but for now I just keep it as it is to make the testing easier with command line option. When the branch is final, I will talk to you about the architecture change.

**Comment by saaramahmoudi** [comment]
> Quite a few CI failures ... can you take a look @saaramahmoudi ?

Yes, I will. But before that, we still have some routing problem that might be related to router lookahead. I already talked to @amin1377, and he agreed to talk a look into the new commits. After we found all bugs, I will investiagte CI issues.

**Comment by vaughnbetz** [comment]
Please go ahead and merge this after the conflicts are resolved, and a QoR test on Titan to ensure 2D QoR isn't affected.
There are two comments above that should go in an issue to resolve afterwards:
- move --custom_3d_sb_fanin_fanout to the architecture file as a new tag, rather than as a command line option. It should go in the <device> section.
- Add a big comment somewhere (maybe in rr_graph.cpp) about how custom_3d_sb_fanin_fanout works.  It should detail exactly what the rr graph this generates looks like.

**Comment by amin1377** [comment]
2D Titan QoR:
image
2D large VTR:

File: vpr/src/route/connection_router.cpp
Comment (vaughnbetz) [inline vpr/src/route/connection_router.cpp:1022]: Need to comment what you are doing here and why.

**Comment by vaughnbetz** [comment]
Looks good. One commenting update requested, plus we should add the missing comments I flagged before. I'll create an issue to track that.

**Pull request merged**
